### ISSUE TITLE
ONNX Training work for ONNX release 1.8 (and maybe beyond)

### ISSUE THREAD

**Issue by sveta-levitan** [onnx/onnx#2886]
This issue will help us to keep track of the work for ONNX Training for release 1.8, and possibly beyond it.

The tasks as Wei-Cheng Chin outlined them in July 7, 2020 working group meeting:

1. Adding Differentiability tags to all Input and Output parts of ONNX operators.
onnx/defs/tensor/defs.cc [@wschin ] PR #2896, #2954
onnx/defs/rnn/defs.cc [Rajeev @rajeevnalawadi] - PR #2892 
onnx/defs/nn/defs.cc [Rajeev @rajeevnalawadi, Marten @ma-hei]  PR #2893 , #3011
onnx/defs/math/defs.cc [Svetlana @sveta-levitan ] - PR #2911 
onnx/defs/logical/defs.cc [Rajeev @rajeevnalawadi #2966]
onnx/defs/reduction/defs.cc [Svetlana @sveta-levitan #3053 ]

2. APIs to create training graphs. [@wschin, #3008]
3. Create samples for training graphs. [@wschin, #3008]
4. Remove GraphCall. [@wschin #2964]

-----------post 1.8 release-----------

5. Update checker to check the use of Gradient based on the tags. [@wschin]
6. Replace optimizers with functions (not sure how to handle variable-length inputs and outputs). [Adagrad @sveta-levitan, Adam, Momentum]. To make this replacement feasible, we may need to change the variadic input/output behavior of those optimizers to fixed-length.
7. Create actually training loop (starting with linear regression) using training spec (evaluating training graph like a inference graph).

**Comment by wschin** [comment]
Item 4) is to check if the use of Gradient is reasonable. To compute gradient of tensor `Y` w.r.t. tensor `X`, we need to make sure the path from `X` to `Y` are differentiable. Also, only members in float-pointing tensor family has gradient; all other types don't have gradient. Given a Gradient operator, the checker should traverse the `differentiated` sub-graph and see if all nodes (operators) are differentiable.

**Comment by askhade** [comment]
Please move the relevant open tasks to track the upcoming 1.10 release or close this issue.
Thanks!

**Comment by askhade** [comment]
@sveta-levitan , @wschin : Can you move the remaining items under 1.10 tracking issue.
